astropy/units/quantity_helper/function_helpers.py

```diff
@@ -987,9 +987,10 @@ def array2string(a, *args, **kwargs):
         a = a.value
     else:
         # See whether it covers our dtype.
-        from numpy.core.arrayprint import _get_format_function
+        from numpy.core.arrayprint import _get_format_function, _make_options_dict
```

Comment:
@mhvk , should we not use hidden functions from deep within numpy? I am afraid further refactoring upstream will come back to bite us here.

Reply:
In principle, yes, and clearly here my initial implementation was even wrong! (since without `_make_options_dict`, the `legacy` option does not get properly translated into an integer). But the alternative is duplication of the whole `arrayprint` module...

Anyway, I hope this will become moot: in the initial scalar work, a public version of `_get_format_function` was included, and I think this is still likely to happen. Once that is there, we can just switch to using it.